### ISSUE TITLE
Permanent tooltips positioned incorrectly on first load

### ISSUE THREAD

**Issue by Techn1x** [miguelcobain/ember-leaflet#165]
https://ember-twiddle.com/92821d0f3817f2565bbb5b470c5bbf35/92f7e4e4017e4d5b9a2633a8b95ba957c835a743?openFiles=controllers.application.js%2C

Open that twiddle, and if you don't zoom in, you can see the tooltips are not positioned correctly (they're directly on top of the markers). Only occurs when `permanent=true` (when it's false, tooltips are always positioned correctly on hover). Sometimes, the tooltips are ok even if `permanent=true`, see the twiddle for a demo of this.

I can't seem to replicate it using standard leaflet.js (see [this jsfiddle](https://jsfiddle.net/ZkC5M/476/)) so I think it's an issue with ember-leaflet...

Let me know if you need any additional information. I did a bit of investigating but couldn't really figure out anything useful

**Comment by Techn1x** [comment]
I may have uncovered something useful... in leaflet.js when the tooltip position is calculated https://github.com/Leaflet/Leaflet/blob/master/src/layer/Tooltip.js#L134-L165

If the tooltip is `permanent=false` and it was triggered with a hover, the container element has a width and height, so these values are correct
```javascript
tooltipWidth = container.offsetWidth,       // eg 85px
tooltipHeight = container.offsetHeight,    // eg 50px
```

If the tooltip is `permanent=true` and it was drawn on load, the container element seems to be empty, so it is its minimum size (width 14, height 14), so these values are wrong
```javascript
tooltipWidth = container.offsetWidth,     // 14px
tooltipHeight = container.offsetHeight,   // 14px
```

These values are then used to calculate the tooltip position.

Because my tooltips are always static size (85px,50px), one quick and dirty workaround is to do this in my CSS...
```sass
.leaflet-tooltip {
    .min-width: 85px;
    .min-height: 50px;
}
```
That way, even when the container is empty, it is at the size it should be with the expected content, so the position calculations are correct.. but this is far from ideal :p

**Comment by TheSegfault** [comment]
Hi, I found another solution for ur problem, I personaly .openTooltip(); and .closeTooltip(); at the initialization.

It look like that 

`              marker.bindTooltip('<img src=' + geoConfig.description.img + '/>' + geoConfig.description.title);
                marker.bindTooltip('<img src=' + geoConfig.description.img + '/>' + geoConfig.description.title).openTooltip();
                marker.bindTooltip('<img src=' + geoConfig.description.img + '/>' + geoConfig.description.title).closeTooltip();`

**Comment by Techn1x** [comment]
Thanks!! I hadn't managed to find a workaround for it yet, so I'm excited to try your idea next chance I get.

Unfortunately I use the ember handlebars templates to create my tooltips so I'm going to need to find a way to reference them so that I can open/close them

**Comment by Techn1x** [comment]
So I've spent a few more hours looking into this now with a clear mind and I can see the cause and solution to the problem.

Looking at the `tooltip-layer.js` file, there are some lines of code that hijack/override the `onAdd` method for the layer, with the specific purpose of rendering the tooltip template *before* the tooltip is opened (so that the tooltip will have it's dimensions set according to the rendered DOM)
https://github.com/miguelcobain/ember-leaflet/blob/1d72052bf391b8af7e58720df25809416b9bda82/addon/components/tooltip-layer.js#L42-L58

This works wonderfully for tooltips where `permanent=false` because I believe the dimensions aren't set until you hover over the marker (and by this point the tooltip is definitely already rendered)

When `permanent=true` however, the layer's original onAdd method is called before the onAdd hijack occurs.

So my fix is to move that hijack code to be earlier in the process.

My tooltip-layer-fixed.js now looks like this;
```javascript
//app/components/leaflet/tooltip-layer-fixed.js
import Ember from 'ember';
import TooltipLayer from 'ember-leaflet/components/tooltip-layer';

export default TooltipLayer.extend({
  _addPopupListeners() {
    // we need to user `layerremove` event becase it's the only one that fires
    // *after* the popup was completely removed from the map
    let parentComponent = this.get('parentComponent');
    parentComponent._layer._map.addEventListener('layerremove', this._onLayerRemove, this);
  },

  addToContainer() {
    this.hijackOnAddMethod();
    this.get('parentComponent')._layer.bindTooltip(this._layer);
  },
	
  hijackOnAddMethod() {
    // we need to hijack the `onAdd` method because we need to
    // render the template *before* the popup is opened.
    // This way, the popup will set its dimensions according to the rendered DOM.
    let oldOnAdd = this._layer.onAdd;
    this._layer.onAdd = (map) => {
      // trigger _initLayout manually, otherwise Tooltip doesn't have the container set
      // to calculate initial position
      if (!this._layer._container) {
        this._layer._initLayout();
      }
      // this will make wormwhole render to the document fragment
      this.set('shouldRender', true);
      // ember-wormhole will render on the afterRender queue, so we need to render after that
      Ember.run.next(() => {
        if (this.get('shouldRender')) {
          oldOnAdd.call(this._layer, map);
        }
      });
    };	
  }
})
```
NOTE: The above code also includes the fix from this PR
https://github.com/miguelcobain/ember-leaflet/pull/138

@miguelcobain is this a good fix? Should I do a PR or is there a better way to achieve this?

Thanks

**Comment by Techn1x** [comment]
Made a PR anyway - you're welcome to modify/close/etc or make suggestions

**Comment by kaermorchen** [comment]
@Techn1x Thank you bro. You have safed my time.

**Comment by Techn1x** [comment]
No worries @kaermorchen , glad I saved someone the many hours/days I spent on this.

I wish I could get some maintainers' feedback on these fixes I've made, feel like I've been talking to a wall this whole time :P Although there does seem to be a lot of good PR's that haven't been acknowledged so I assume it's a time issue with the maintainers.

**Comment by miguelcobain** [comment]
@Techn1x seems that with ember-leaflet 3.0.18 the twiddle no longer fails.

Here is the updated twiddle: https://ember-twiddle.com/00873a8d886530de46f22c85970f9360?openFiles=twiddle.json%2C

I'm closing this and #186 
Thank you.

**Comment by Techn1x** [comment]
Thanks! That's fantastic news, can't wait to update ember-leaflet in my project first thing Monday!

**Comment by Techn1x** [comment]
Unfortunately, even though the ember-twiddle does seem to show its fixed in 3.0.18+, I'm still having the same issue after updating to 3.1.2 and removing my workarounds.

Stepping through the code, it seems to call `addToContainer` before `didCreateLayer`, meaning the tooltip gets binded before `_addPopupListeners` (ie the onAdd method is overridden)

I can confirm the workaround is a lot simpler now though ^_^
```javascript
import TooltipLayer from 'ember-leaflet/components/tooltip-layer';
export default TooltipLayer.extend({
	addToContainer() {
		this._addPopupListeners(); // overrides the onAdd method
		this._super(); // binds the tooltip to the layer
	}
})
```

**Comment by luciodale** [comment]
If you are using `getBounds()` to center the shapes, remember to add the tooltips afterwards
